packages/dates/src/format.ts

```diff
@@ -0,0 +1,167 @@
+  };
+};
+
+const TOKEN_MAP: TokenDateTimeFormatOptionsMap = {
+  YYYY: {type: 'year', options: {year: 'numeric'}},
+  YY: {type: 'year', options: {year: '2-digit'}},
+  MMMM: {type: 'month', options: {month: 'long'}},
+  MMM: {type: 'month', options: {month: 'short'}},
+  MM: {type: 'month', options: {month: '2-digit'}},
+  // eslint-disable-next-line id-length
+  M: {type: 'month', options: {month: 'numeric'}},
+  DD: {type: 'day', options: {day: '2-digit'}},
+  // eslint-disable-next-line id-length
+  D: {type: 'day', options: {day: 'numeric'}},
+  HH: {type: 'hour', options: {hour: '2-digit', hour12: false}},
+  // eslint-disable-next-line id-length
+  H: {
+    type: 'hour',
+    options: {hour: 'numeric', hour12: false},
+    // extraParser is needed when Node and Browsers don't respect the 2-digit
+    // and numeric options, which happens occasionally for time units
+    extraParser: numericFromTwoDigit,
+  },
+  hh: {type: 'hour', options: {hour: '2-digit', hour12: true}},
+  // eslint-disable-next-line id-length
+  h: {type: 'hour', options: {hour: 'numeric', hour12: true}},
+  mm: {
+    type: 'minute',
+    options: {minute: '2-digit'},
+    // see 'H' token parsing for more details about extraParser
+    extraParser: twoDigitFromNumeric,
+  },
+  // eslint-disable-next-line id-length
+  m: {type: 'minute', options: {minute: 'numeric'}},
+  ss: {
+    type: 'second',
+    options: {second: '2-digit'},
+    extraParser: twoDigitFromNumeric,
+  },
+  // eslint-disable-next-line id-length
+  s: {type: 'second', options: {second: 'numeric'}},
+  // eslint-disable-next-line id-length
+  A: {
+    type: 'dayPeriod',
+    options: {
+      hour: '2-digit',
+      minute: '2-digit',
+      second: '2-digit',
+      hour12: true,
+    },
+  },
+  // eslint-disable-next-line id-length
+  a: {
+    type: 'dayPeriod',
+    options: {
+      hour: '2-digit',
+      minute: '2-digit',
+      second: '2-digit',
+      hour12: true,
+    },
+    extraParser: lowercase,
+  },
+  dddd: {type: 'weekday', options: {weekday: 'long'}},
+  ddd: {type: 'weekday', options: {weekday: 'short'}},
+};
+
+function numericFromTwoDigit(time: string) {
+  return time.startsWith('0') && time.length === 2 ? time.charAt(1) : time;
+}
+
+function twoDigitFromNumeric(time: string) {
+  return time.length === 1 ? `0${time}` : time;
+}
+
+function lowercase(time: string) {
+  return time.toLowerCase();
+}
+
+const ALL_TOKENS_REGEX = new RegExp(
+  Object.keys(TOKEN_MAP)
+    .map((ch) => `(${ch})`)
+    .join('|'),
+  'g',
+);
+
+export function format(
+  date: Date,
+  fmtStr: string,
+  tz?: string,
+  localeStr?: string,
+): string {
+  const {locale, timeZone} = memoizedGetDateTimeFormat(localeStr, {
+    timeZone: tz,
+  }).resolvedOptions();
+
+  return fmtStr.replace(ALL_TOKENS_REGEX, (token: Token) => {
+    const timeUnit = getTimeUnitFromToken(token, date, locale, timeZone);
+    return timeUnit ?? '';
+  });
+}
+
+function getTimeUnitFromToken(
```

Comment:
This should be inlined with the above function to make things more efficient -- esp. since it's only used by the function above, and also then the above function is not doing a lot on its own :)